### DISCUSSION TITLE
Mark obligatory (non-null) table fields in ERD?

### DISCUSSION THREAD

**Discussion by torstenberg** [mermaid-js/mermaid#3718] [Q&A]
Hi,

is there a specific way to visually mark the table fields in an ER diagram which can't be NULL, i.e. the ones that are obligatory? Sometimes you see them marked with an asterisk (I could add it behind the field name) or they are shown in a bold font (is that possible?). Other times this information is displayed in a separate column. The 'KEY' attribute can only hold PK or FK and this is also a place where this information could be should if Mermaid allowed other text than these two key types.

Any help or insight on this?

**Comment by torstenberg** [comment]
I investigated this further and it seems it is not even possible to have a column name such as `mycolumn*` where the asterisk marks the column as being obligatory (NOT NULL). This renders the diagram invalid in the live editor. So, I am out of options. Any ideas?

**Reply by OsirisDBA** [reply]
Best I've come up with is to just add them to the comment field.

**Reply by kendallroth** [reply]
Would it be possible to relax the attribute type character restrictions (or maybe also attribute names?), allowing for the use of `*` (required) and `?` (optional)? While these would have no functionality other than display purposes, that may be sufficient for a large amount of uses cases. Supporting **both** `*` and `?` would also support two ways of declaring which fields are required/optional, depending on whether people wish to indicate required or optional fields. Additionally, it seems like it _could_ be a relatively simple fix, _as long as_ allowing those characters in types/attribute names doesn't cause other parsing issues (other than being disallowed currently)?

**Comment by shashank-vangari** [comment]
I would like to have this as well ?

**Comment by SleepWalker** [comment]
There is also an issue regarding this topic: https://github.com/mermaid-js/mermaid/issues/4429

**Comment by UnoYakshi** [comment]
I, too, find it a quite useful feature. And relatively easy to implement as it's basically "PK"/"FK" with a different label. It also doesn't bring any real background logic dependencies, because it's all contained within an isolated field/column. Until somebody decides to turn Mermaid-ERD into a full-scale DB schema generation software, that is…

Update: I didn't realize [OsirisDBA's comment](https://github.com/orgs/mermaid-js/discussions/3718#discussioncomment-6222401) was about Mermaid-ERD comments like in the diagram below.
My biggest issue with that is you'd need to:
- either wrap everything in `".."` (rewrites!) and not have comments section
- or have `NN` in the separate field (like in the diagram below)

```mermaid
erDiagram
    User {
        UUID4 id PK
        str email UK "NN"
        text phone UK "or you might decide not to use NN for UK"
        datetime created_at "NN"
        datetime last_modified "NN"
    }
```
(I don't think you need to mark PK as NN. It's kind of implied. Not so sure about UK, though, but I'd personally imply that as well.)

**Comment by UnoYakshi** [comment]
I still think it'd be a good addition. So I'll try to make a PR for that this month. But I'm in no way a TS/JS-developer.

**Comment by mwaddell** [comment]
Allowing "?" in the type column would allow differentiating between `bool` and `bool?` etc
